Reduce use of std::span::subspan() by leveraging ParsingUtilities

cdumez opened the pull request:
#### 7a78d638b1637b95deaa1750b82e1dc0ca9b7d2d
<pre>
Reduce use of std::span::subspan() by leveraging ParsingUtilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=285271">https://bugs.webkit.org/show_bug.cgi?id=285271</a>

Reviewed by Darin Adler.

A lot more code can be ported but this is a step in this direction.

* Source/JavaScriptCore/API/JSScript.mm:
(-[JSScript readCache]):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::CallArguments::CallArguments):
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Encoder::release):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::overrideLanguageScriptRegion):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::canonicalizeUnicodeExtensionsAfterICULocaleCanonicalization):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::jsBinaryIntegerLiteral):
(JSC::jsOctalIntegerLiteral):
(JSC::jsHexIntegerLiteral):
(JSC::toDouble):
(JSC::parseFloat):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::bufferMode&gt;::append):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToDataWithOneCharacterSeparatorRepeatedly):
* Source/JavaScriptCore/runtime/ParseInt.h:
(JSC::isStrWhiteSpace):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/DateMath.cpp:
(WTF::parseES5DatePortion):
(WTF::parseES5TimePortion):
(WTF::parseDate):
* Source/WTF/wtf/text/ParsingUtilities.h: Renamed from Source/WebCore/html/parser/ParsingUtilities.h.
(WTF::isNotASCIISpace):
(WTF::skipExactly):
(WTF::characterPredicate):
(WTF::skipUntil):
(WTF::skipWhile):
(WTF::skipExactlyIgnoringASCIICase):
(WTF::skipLettersExactlyIgnoringASCIICase):
(WTF::skipCharactersExactly):
(WTF::consumeSingleElement):
(WTF::consumeSpan):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readString):
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/html/HTMLFontElement.cpp:
(WebCore::parseFontSize):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::parseHTMLIntegerInternal):
(WebCore::isHTMLSpaceOrDelimiter):
(WebCore::parseHTMLListOfOfFloatingPointNumberValuesInternal):
(WebCore::parseHTTPRefreshInternal):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
* Source/WebCore/html/track/VTTScanner.h:
(WebCore::characterPredicate):
(WebCore::VTTScanner::characterPredicate): Deleted.
* Source/WebCore/loader/LinkHeader.cpp:
* Source/WebCore/loader/ResourceCryptographicDigest.cpp:
* Source/WebCore/loader/SubresourceIntegrity.cpp:
* Source/WebCore/loader/appcache/ApplicationCacheManifestParser.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicyMediaListDirective.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.cpp:
* Source/WebCore/platform/DateComponents.cpp:
* Source/WebCore/platform/network/RFC8941.cpp:
* Source/WebCore/svg/SVGParserUtilities.h:

Canonical link: <a href="https://commits.webkit.org/288353@main">https://commits.webkit.org/288353@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc07060094f36f1b7918381e4dabc2632c1cb10d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36962 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87901 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33840 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84875 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10346 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64500 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22266 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85838 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75297 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44776 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29485 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32873 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75767 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72996 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30168 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89268 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81830 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10074 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7313 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72917 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71105 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72132 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16294 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12816 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10027 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15548 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104235 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9901 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25262 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13366 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11670 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
```diff
@@ -99,8 +99,8 @@ JSC_DEFINE_HOST_FUNCTION(uint8ArrayConstructorFromBase64, (JSGlobalObject* globa
-    uint8_t* data = uint8Array->typedVector();
-    memcpySpan(std::span { data, data + writeLength }, output.span().subspan(0, writeLength));
+    auto data = uint8Array->typedSpan();
+    memcpySpan(data, output.span().first(writeLength));
```
Might read better without the local variable.

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -3502,10 +3503,9 @@ class CloneDeserializer : public CloneBase {
             if (span.size() < length)
                 return false;
```
Could use `consumeSpan` to do this length check to and put the `span` into a local variable.

cdumez replied:
`consumeSpan()`, as proposed, doesn't do a length check (it would RELEASE_ASSERT). I guess we could make it do a length check and return an empty span, then leave it to the caller to check that they got a span for the right size?

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -396,38 +387,29 @@ static bool parseHTTPRefreshInternal(std::span<const CharacterType> data, double
-        quote = data.front();
-        data = data.subspan(1);
+        quote = consumeSingleElement(data);
     } else
```
Take off the braces too? I also considered if it would read better using `skipExactly` and it _almost_ does.

cdumez commented on Source/WTF/wtf/text/ParsingUtilities.h:
```diff
@@ -255,6 +290,37 @@ template<typename CharacterType, std::size_t Extent> constexpr bool skipCharacte
+    return value;
+}
+
+template<typename T> std::span<T> consumeSpan(std::span<T>& data, size_t amountToConsume)
```
@darinadler Let me know if you'd prefer if this function (and `consumeSingleElement`) do a length check. Or maybe we'd add another `tryConsumeSpan` that does a length check. For now I kept as is.

cdumez committed:
Reduce use of std::span::subspan() by leveraging ParsingUtilities
https://bugs.webkit.org/show_bug.cgi?id=285271

Reviewed by Darin Adler.

A lot more code can be ported but this is a step in this direction.

* Source/JavaScriptCore/API/JSScript.mm:
(-[JSScript readCache]):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::CallArguments::CallArguments):
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Encoder::release):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::overrideLanguageScriptRegion):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::canonicalizeUnicodeExtensionsAfterICULocaleCanonicalization):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::jsBinaryIntegerLiteral):
(JSC::jsOctalIntegerLiteral):
(JSC::jsHexIntegerLiteral):
(JSC::toDouble):
(JSC::parseFloat):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::bufferMode>::append):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToDataWithOneCharacterSeparatorRepeatedly):
* Source/JavaScriptCore/runtime/ParseInt.h:
(JSC::isStrWhiteSpace):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/DateMath.cpp:
(WTF::parseES5DatePortion):
(WTF::parseES5TimePortion):
(WTF::parseDate):
* Source/WTF/wtf/text/ParsingUtilities.h: Renamed from Source/WebCore/html/parser/ParsingUtilities.h.
(WTF::isNotASCIISpace):
(WTF::skipExactly):
(WTF::characterPredicate):
(WTF::skipUntil):
(WTF::skipWhile):
(WTF::skipExactlyIgnoringASCIICase):
(WTF::skipLettersExactlyIgnoringASCIICase):
(WTF::skipCharactersExactly):
(WTF::consumeSingleElement):
(WTF::consumeSpan):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readString):
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/html/HTMLFontElement.cpp:
(WebCore::parseFontSize):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::parseHTMLIntegerInternal):
(WebCore::isHTMLSpaceOrDelimiter):
(WebCore::parseHTMLListOfOfFloatingPointNumberValuesInternal):
(WebCore::parseHTTPRefreshInternal):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
* Source/WebCore/html/track/VTTScanner.h:
(WebCore::characterPredicate):
(WebCore::VTTScanner::characterPredicate): Deleted.
* Source/WebCore/loader/LinkHeader.cpp:
* Source/WebCore/loader/ResourceCryptographicDigest.cpp:
* Source/WebCore/loader/SubresourceIntegrity.cpp:
* Source/WebCore/loader/appcache/ApplicationCacheManifestParser.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicyMediaListDirective.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
* Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.cpp:
* Source/WebCore/platform/DateComponents.cpp:
* Source/WebCore/platform/network/RFC8941.cpp:
* Source/WebCore/svg/SVGParserUtilities.h:

Canonical link: https://commits.webkit.org/288353@main
